django/contrib/postgres/constraints.py

```diff
@@ -0,0 +1,85 @@
+from django.db.models.constraints import BaseConstraint
+from django.db.models.sql import Query
+
+__all__ = ['ExclusionConstraint']
+
+
+class ExclusionConstraint(BaseConstraint):
+    template = 'CONSTRAINT %(name)s EXCLUDE USING %(index_type)s(%(expressions)s)%(where)s'
+
+    def __init__(self, *, expressions=(), index_type=None, condition=None, name=''):
+        if index_type and index_type.lower() not in {'gist', 'spgist'}:
+            raise ValueError('Exclusion constraints only support GiST or SP-GiST indexes.')
+        if not expressions:
+            raise ValueError('The expressions argument is mandatory.')
+        if condition is not None and not isinstance(condition, Q):
+            raise ValueError('The condition argument needs to be either None or a Q-instance.')
+        self.expressions = expressions
+        self.index_type = index_type
+        self.condition = condition
+        super().__init__(name=name)
+
+    def _parse_expressions(self, compiler, connection, query):
+        expressions = []
+        for expression, _ in self.expressions:
+            if isinstance(expression, (F, str)):
+                base = F(expression) if isinstance(expression, str) else expression
+                expression = base.resolve_expression(query=query, simple_col=True)
+            expression = expression.resolve_expression(query=query)
+            sql, params = expression.as_sql(compiler, connection)
+            expressions.append(sql % params)
+        return [
+            '%s WITH %s' % (expression, self.expressions[idx][1])
+            for idx, expression in enumerate(expressions)
+        ]
+
+    def _parse_where(self, compiler, schema_editor, query):
+        if self.condition is None:
+            return ''
+        where = query.build_where(self.condition)
+        sql, params = where.as_sql(compiler, schema_editor.connection)
+        params = tuple(schema_editor.quote_value(p) for p in params)
+        return ' WHERE ({})'.format(sql % params)
+
+    def constraint_sql(self, model, schema_editor):
+        query = Query(model)
+        compiler = schema_editor.connection.ops.compiler('SQLCompiler')(query, schema_editor.connection, 'default')
+        expressions = self._parse_expressions(compiler, schema_editor.connection, query)
+        where = self._parse_where(compiler, schema_editor, query)
+        return self.template % {
+            'name': schema_editor.quote_name(self.name),
+            'index_type': self.index_type or 'GIST',
+            'expressions': ', '.join(expressions),
+            'where': where,
+        }
+
+    def deconstruct(self):
+        kwargs = {
+            'expressions': self.expressions,
+            'name': self.name,
+        }
+        if self.condition is not None:
+            kwargs['condition'] = self.condition
+        if self.index_type is not None:
+            kwargs['index_type'] = self.index_type
+        return self.__class__.__qualname__, (), kwargs
```

Comment:
Why we don't use `super().deconstruct()`?
```python
def deconstruct(self):
    path, args, kwargs = super().deconstruct()
    kwargs['expressions'] = self.expressions
    if self.condition is not None:
        kwargs['condition'] = self.condition
    if self.index_type != 'GIST':
        kwargs['index_type'] = self.index_type
    return path, args, kwargs
```

Reply:
Also here:
```python
if self.index_type.lower() != 'gist':
```

Reply:
I think because `path` will then be `django.db.models` without additional work.

Reply:
It will be `django.contrib.postgres.constraints.XXX` but I don't think that's an issue for core constraints we return `django.db.models.XXX`.

Reply:
I just realiased :)